internal/multistorage/cache/status_cache_test.go

```diff
@@ -0,0 +1,280 @@
+package cache
+
+import (
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/wal-g/wal-g/pkg/storages/memory"
+	"github.com/wal-g/wal-g/pkg/storages/storage"
+)
+
+func TestStatusCache(t *testing.T) {
+	// Run these tests sequentially because they share common resources
+	t.Run("all alive storages", testAllAliveStorages)
+	t.Run("first alive storage", testFirstAliveStorage)
+	t.Run("specific storage", testSpecificStorage)
+}
+
+func testAllAliveStorages(t *testing.T) {
+	initTest(t)
+
+	cache := newTestCache(2)
+
+	t.Run("check for alive if not cached", func(_ *testing.T) {
+		alive, err := cache.AllAliveStorages()
+		require.NoError(t, err)
+		require.Len(t, alive, 3)
+		require.Equal(t, "default", alive[0].Name)
+		require.Equal(t, "failover_1", alive[1].Name)
+		require.Equal(t, "failover_2", alive[2].Name)
+	})
+
+	t.Run("create new cache file", func(_ *testing.T) {
+		statuses, err := readFile()
+		require.NoError(t, err)
+		require.Len(t, statuses, 3)
+	})
+
+	updateInMem("failover_2", false)
+
+	t.Run("take statuses from memory if relevant", func(_ *testing.T) {
+		alive, err := cache.AllAliveStorages()
+		require.NoError(t, err)
+		require.Len(t, alive, 2)
+		require.Equal(t, "default", alive[0].Name)
+		require.Equal(t, "failover_1", alive[1].Name)
+	})
+
+	invalidateInMem("default")
+	updateInFile(t, "failover_1", false)
+
+	t.Run("take statuses from file if it exists and memory is outdated", func(_ *testing.T) {
+		alive, err := cache.AllAliveStorages()
+		require.NoError(t, err)
+		require.Len(t, alive, 2)
+		require.Equal(t, "default", alive[0].Name)
+		require.Equal(t, "failover_2", alive[1].Name)
+	})
+
+	invalidateInMem("default")
+	invalidateInFile(t, "failover_2")
+
+	t.Run("check missing storages for alive and take others from file", func(_ *testing.T) {
+		alive, err := cache.AllAliveStorages()
+		require.NoError(t, err)
+		require.Len(t, alive, 2)
+		require.Equal(t, "default", alive[0].Name)
+		require.Equal(t, "failover_2", alive[1].Name)
+	})
+
+	t.Run("update existing cache file", func(_ *testing.T) {
+		statuses, err := readFile()
+		require.NoError(t, err)
+		require.True(t, statuses.isRelevant(time.Hour, cache.storagesInOrder[2]))
+	})
+
+	invalidateInMem("default")
```

Comment:
What happens if all storages in cache are marked as failed?

Reply:
It would be an error. I added such test case